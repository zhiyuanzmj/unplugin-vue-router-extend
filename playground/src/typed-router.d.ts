/* eslint-disable */
/* prettier-ignore */
// @ts-nocheck
// Generated by unplugin-vue-router. ‼️ DO NOT MODIFY THIS FILE ‼️
// It's recommended to commit this file.
// Make sure to add this file to your tsconfig.json file as an "includes" or "files" entry.

declare module 'vue-router/auto-routes' {
  import type {
    RouteRecordInfo,
    ParamValue,
    ParamValueOneOrMore,
    ParamValueZeroOrMore,
    ParamValueZeroOrOne,
  } from 'vue-router'

  /**
   * Route name map generated by unplugin-vue-router
   */
  export interface RouteNamedMap {
    'index': RouteRecordInfo<'index', '/', Record<never, never>, Record<never, never>>,
    'all': RouteRecordInfo<'all', '/:all(.*)', { all: ParamValue<true> }, { all: ParamValue<false> }>,
    'sensor': RouteRecordInfo<'sensor', '/:sensor', { sensor: ParamValue<true> }, { sensor: ParamValue<false> }>,
    'sensor-current': RouteRecordInfo<'sensor-current', '/:sensor/current', { sensor: ParamValue<true> }, { sensor: ParamValue<false> }>,
    'about/': RouteRecordInfo<'about/', '/about', Record<never, never>, Record<never, never>>,
    'about': RouteRecordInfo<'about', '/about', Record<never, never>, Record<never, never>>,
    'about-user-id': RouteRecordInfo<'about-user-id', '/about/:id', { id: ParamValue<true> }, { id: ParamValue<false> }>,
    'about-id-more': RouteRecordInfo<'about-id-more', '/about/:id/more', { id: ParamValue<true> }, { id: ParamValue<false> }>,
    'about-id-nested': RouteRecordInfo<'about-id-nested', '/about/:id/nested', { id: ParamValue<true> }, { id: ParamValue<false> }>,
    'blog': RouteRecordInfo<'blog', '/blog', Record<never, never>, Record<never, never>>,
    'blog-id': RouteRecordInfo<'blog-id', '/blog/:id', { id: ParamValue<true> }, { id: ParamValue<false> }>,
    'blog-today': RouteRecordInfo<'blog-today', '/blog/today', Record<never, never>, Record<never, never>>,
    'blog-today-all': RouteRecordInfo<'blog-today-all', '/blog/today/:all(.*)', { all: ParamValue<true> }, { all: ParamValue<false> }>,
    'components': RouteRecordInfo<'components', '/components', Record<never, never>, Record<never, never>>,
  }
}
