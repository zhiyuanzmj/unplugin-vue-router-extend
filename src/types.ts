import type { TreeNode } from 'unplugin-vue-router'

export interface Options {
  // define your plugin options here
  routeMap: Map<string, TreeNode>
}
