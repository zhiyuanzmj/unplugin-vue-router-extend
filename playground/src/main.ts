import { createApp } from 'vue'
import { createRouter, createWebHistory } from 'vue-router/auto'
import App from './App.vue'

const router = createRouter({
  history: createWebHistory(),
})

const app = createApp(App)

app.use(router)

app.mount('#app')
