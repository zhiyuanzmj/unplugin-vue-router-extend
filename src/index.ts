import type { Program } from '@babel/types'
import type { Options } from './types'
import { addNormalScript, babelParse, generateTransform, getLang, MagicString, parseSFC } from '@vue-macros/common'
import { createUnplugin } from 'unplugin'

export * from './utils'

export default createUnplugin<Options>(options => ({
  name: 'unplugin-vue-router-extend',
  enforce: 'pre',
  transform(code, id) {
    const name = options.routeMap.get(id)?.name
    if (!name)
      return

    const lang = getLang(id)
    const s = new MagicString(code)
    if (lang === 'vue') {
      const sfc = parseSFC(code, id)
      if (sfc.script) {
        const offset = sfc.script.loc.start.offset
        transformAST(sfc.getScriptAst()!, name, s, offset)
      }
      else {
        const normalScript = addNormalScript(sfc, s)
        const scriptOffset = normalScript.start()
        s.appendLeft(
          scriptOffset,
          `\nimport { defineComponent } from 'vue'
export default /* @__PURE__ */ defineComponent({
  name: '${name}'
})`,
        )
        normalScript.end()
      }
    }
    else if (/[jt]sx?$/.test(lang)) {
      const ast = babelParse(code, lang)
      transformAST(ast, name, s)
    }

    return generateTransform(s, id)
  },
}))

function transformAST(ast: Program, name: string, s: MagicString, offset = 0) {
  for (const stmt of ast.body) {
    if (
      stmt.type === 'ExportDefaultDeclaration'
      && stmt.declaration.type === 'CallExpression'
      && stmt.declaration.callee.type === 'Identifier'
      && stmt.declaration.callee.name === 'defineComponent'
    ) {
      const argument = stmt.declaration.arguments.find(i => i.type === 'ObjectExpression')
      if (argument) {
        if (!argument.properties?.find(
          prop =>
            prop.type === 'ObjectProperty'
            && prop.key.type === 'Identifier'
            && prop.key.name === 'name',
        )) {
          s.appendLeft(
            offset + argument.end! - 1,
            `${
              !argument.extra?.trailingComma && argument.properties.length
                ? ','
                : ''
            } name: '${name}'`,
          )
        }
      }
      else {
        s.appendRight(offset + stmt.declaration.end! - 1, `, { name: '${name}' }`)
      }
    }
  }
}
