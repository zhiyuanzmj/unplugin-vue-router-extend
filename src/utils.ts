import { getFileBasedRouteName } from 'unplugin-vue-router'
import type { TreeNode } from 'unplugin-vue-router'
import { debounce } from 'lodash-unified'

export function getNuxtStyleRouteName(node: TreeNode): string {
  if (node.parent?.isRoot() && node.value.pathSegment === '')
    return 'index'

  let name = node.value.subSegments
    .map((segment) => {
      if (typeof segment === 'string')
        return `-${segment}`

      // else it's a param
      return segment.paramName
    })
    .join('')

  if (node.value.filePaths.size && node.children.has('index'))
    name += '/'

  const parent = node.parent
  return (
    (parent && !parent.isRoot()
      ? getNuxtStyleRouteName(parent).replace(/\/$/, '')
      : '') + name
  ).replace(/^-|-$/, '')
}

function _getRouteMap(node: TreeNode, { routeMap, nuxtStyle }: {
  routeMap: Map<string, TreeNode>
  nuxtStyle: boolean
}) {
  node.children.forEach((i) => {
    if (i.value.filePaths.get('default'))
      routeMap.set(i.value.filePaths.get('default')!, i)

    // else if (nuxtStyle) {
    //   i.parent?.children.delete(i.value.rawSegment)
    //   i.children.forEach((child) => {
    //     const block: any = {
    //       path: i.path + (child.path ? `/${child.path}` : ''),
    //     }
    //     if (i.value.isParam() && !child.path)
    //       block.props = true
    //     child.setCustomRouteBlock(child.path, block)
    //     i.parent?.children.set(child.value.rawSegment || i.value.rawSegment, child)
    //   })
    // }

    if (i.children.size)
      _getRouteMap(i, { routeMap, nuxtStyle })
  })
}

function getRoot(node: TreeNode): TreeNode {
  return node.parent ? getRoot(node.parent) : node
}

const getRouteMap = debounce(
  (node, options) => _getRouteMap(getRoot(node), options),
  100,
  { leading: true, trailing: false },
)

export const getRouteName = ({
  routeMap = new Map<string, TreeNode>(),
  nuxtStyle = false,
}, getRouteName = nuxtStyle ? getNuxtStyleRouteName : getFileBasedRouteName) =>
  (node: TreeNode) => {
    getRouteMap(node, { routeMap, nuxtStyle })

    return getRouteName(node)
  }
