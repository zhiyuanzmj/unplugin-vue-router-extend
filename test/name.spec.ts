import { resolve } from 'node:path'
import { describe, expect, it } from 'vitest'
import glob from 'fast-glob'
import { build } from 'vite'
import VueRouter from 'unplugin-vue-router/vite'
import vueRouterExtend from '../src/vite'
import { getRouteName } from '../src'

describe('generate component name', async () => {
  const root = resolve(__dirname, '../playground/src/pages')

  const files = await glob('./**/*.{vue,js,ts}', {
    cwd: root,
    onlyFiles: true,
  })

  for (const file of files) {
    const routeMap = new Map()
    it(file.replace(/\\/g, '/'), async () => {
      const filepath = resolve(root, file)

      const bundle = await build({
        configFile: false,
        build: {
          sourcemap: false,
          write: false,
          lib: {
            entry: filepath,
            formats: ['es'],
            name: 'index',
          },
        },
        plugins: [
          VueRouter({
            routesFolder: root,
            getRouteName: getRouteName({ routeMap, nuxtStyle: true }),
            dts: resolve(root, '../typed-router.d.ts'),
            exclude: ['**/__test__/**'],
          }),
          vueRouterExtend({
            routeMap,
          }),
          {
            name: 'to-string',
            transform: code =>
              `export default \`${code.replace(/`/g, '\\`')}\``,
          },
        ],
      })
      if (!Array.isArray(bundle))
        return
      const code = bundle[0].output
        .map(file => (file.type === 'chunk' ? file.code : file.fileName))
        .join('\n')
      expect(code).toMatchSnapshot()
    })
  }
})
