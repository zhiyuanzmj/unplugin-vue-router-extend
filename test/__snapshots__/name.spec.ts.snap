// Vitest Snapshot v1

exports[`generate component name > [...all].vue 1`] = `
"const e = \`<script>
import { defineComponent } from 'vue'
export default /*#__PURE__*/ defineComponent({
  name: 'all'
})
<\\\\/script>
<template>
  <div>...all route</div>
</template>
\`;
export {
  e as default
};
"
`;

exports[`generate component name > [sensor].vue 1`] = `
"const e = \`<template>
  nested dynamic view:
  <router-view />
</template>
\`;
export {
  e as default
};
"
`;

exports[`generate component name > [sensor]/current.vue 1`] = `
"const e = \`<template>
  <p>/[sensor]/current.vue</p>
</template>
\`;
export {
  e as default
};
"
`;

exports[`generate component name > __test__/index.vue 1`] = `
"const e = \`<template>
  test default exclude
</template>
\`;
export {
  e as default
};
"
`;

exports[`generate component name > about.vue 1`] = `
"const e = \`<template>
  <div>
    nested about view:
    <router-view />
  </div>
</template>

<route lang=\\"yml\\">
meta:
  lang: yml
</route>
\`;
export {
  e as default
};
"
`;

exports[`generate component name > about/[id].vue 1`] = `
"const e = \`<script setup lang=\\"ts\\">
defineProps({
  id: String,
})
<\\\\/script>

<template>
  <div>
    <div>
      about/[id].vue: {{ id }}
    </div>
    <router-link to=\\"/about/1b234bk12b3/more\\">
      more deep
    </router-link>
    <router-view />
  </div>
</template>

<route>
{
  name: 'about-user-id',
  meta: {
    requiresAuth: true,
  },
}
</route>
\`;
export {
  e as default
};
"
`;

exports[`generate component name > about/[id]/more.vue 1`] = `
"const e = \`<template>
  <div class=\\"deep-more\\">
    deep nested: about/[id]/more.vue
  </div>
</template>
\`;
export {
  e as default
};
"
`;

exports[`generate component name > about/[id]/nested.vue 1`] = `
"const e = \`<template>
  <div class=\\"deep-nested\\">
    deep nested: about/[id]/nested.vue
  </div>
</template>
\`;
export {
  e as default
};
"
`;

exports[`generate component name > about/index.vue 1`] = `
"const t = \`<template>
  <div>
    about/index.vue
    <router-link to=\\"/about/1b234bk12b3\\">
      id: 1b234bk12b3
    </router-link>
  </div>
</template>
\`;
export {
  t as default
};
"
`;

exports[`generate component name > blog/[id].vue 1`] = `
"const e = \`<script setup lang=\\"tsx\\">
defineProps({
  id: String,
})
<\\\\/script>

<template>
  <div>
    <p>blog/[id].vue: {{ id }}</p>
    <p>
      {{ $route }}
    </p>
  </div>
</template>

<route>
{
  \\"name\\": \\"blog-id\\",
  \\"meta\\":{
    \\"requiresAuth\\": false
  }
}
</route>
\`;
export {
  e as default
};
"
`;

exports[`generate component name > blog/index.vue 1`] = `
"const o = \`<template>
  <div>
    blog/index.vue
    <router-link to=\\"/blog/1b234bk12b3\\">
      id: 1b234bk12b3
    </router-link> |
    <router-link to=\\"/blog/today\\">
      today
    </router-link> |
    <router-link to=\\"/blog/today/xxx\\">
      child - not found
    </router-link>
  </div>
</template>
\`;
export {
  o as default
};
"
`;

exports[`generate component name > blog/today/[...all].vue 1`] = `
"const t = \`<template>
  <div>blog/today ...all route</div>
</template>
\`;
export {
  t as default
};
"
`;

exports[`generate component name > blog/today/index.vue 1`] = `
"const e = \`<template>
  <div>
    blog/today/index.vue
  </div>
</template>
\`;
export {
  e as default
};
"
`;

exports[`generate component name > components.vue 1`] = `
"const e = \`<template>
  <h1>Test Components name</h1>
</template>

<route lang=\\"yaml\\">
meta:
  lang: yaml
</route>
\`;
export {
  e as default
};
"
`;

exports[`generate component name > index.vue 1`] = `
"const r = \`<template>
  <div>
    <p>index.vue</p>
    <router-link to=\\"/blog\\">
      blog
    </router-link> |
    <router-link to=\\"/about\\">
      about
    </router-link> |
    <router-link to=\\"/components\\">
      components
    </router-link> |
    <router-link to=\\"/markdown\\">
      markdown
    </router-link> |
    <router-link to=\\"/xxx/xxx\\">
      not exists
    </router-link> |
    <router-link to=\\"/features/dashboard\\">
      features:dashboard
    </router-link> |
    <router-link to=\\"/features/admin\\">
      features:admin
    </router-link> |
    <router-link to=\\"/admin\\">
      admin
    </router-link>
  </div>
</template>

<route lang=\\"json\\">
{
  \\"meta\\": {
    \\"requiresAuth\\": false
  }
}
</route>
\`;
export {
  r as default
};
"
`;

exports[`generate component name > typed-router.d.ts 1`] = `
"const t = \\"\\";
export {
  t as default
};
"
`;
